### ISSUE TITLE
remove @Nullable from LoadingCache.get()

### ISSUE THREAD

**Issue by cvmocanu** [ben-manes/caffeine#594]
Currently `LoadingCache.get()` is marked with `@Nullable`.
This is incorrect, because Caffeine can't know the nullability of the cache loader function.

This causes a lot of trouble in actual usage, when you have a cache loader that never returns null (e.g. it's a simple `new MyClass()` or some other factory call that never returns null).
Because of the `@Nullable` on the `LoadingCache.get()` method, IntelliJ marks all usages of the result of the `get()` method as potential NPEs.
As a result, we either have to sprinkle the code with asserts or with suppress-warnings, neither of which is really required, because the `get()` method will never return null.

The annotation should be removed, because Caffeine can't know if the result is nullable or not.

**Comment by ben-manes** [comment]
What is the correct way to mark this method to satisfy IntelliJ? If we remove the annotation then the defaults from package-info is that it is non-null, which may be incorrect.  

https://github.com/ben-manes/caffeine/blob/15330a16b509b56ae32a9bd0cb703e36ff6caa1f/guava/src/main/java/com/github/benmanes/caffeine/guava/package-info.java#L24

I don't think there is a correct way to annotate this method, since unlike `@PolyNull` on `get(key, mappingFunction)` we can't peek into the `CacheLoader`. I am fine with doing the least bad option, though. What would [JSpecify](https://github.com/jspecify/jspecify) recommend in this case  @cpovirk?

/cc @DavyLandman @vlsi

**Comment by vlsi** [comment]
1. `Map#get` is often quoted when arguing on null safety. I think `@Nullable ... get()` is the best we can have with the current tooling.

    See the relevant discussions in https://github.com/jspecify/jspecify/issues/2

2. However, `LoadingCache#get()` seems to be a bit different: `LoadingCache#get()` seems to return `null` if and only if the loader returns `null`, so if the loader return type is non-nullable, then `LoadingCache#get()` should be non-nullable.  So the following might be slightly better (I don't remember if that signature has been discussed earlier):

```java
public interface LoadingCache<K extends Object, V extends @Nullable Object> extends Cache<K, V> {
  V get(K key);
```

    What it means is `V` can be any type, including `nullable` or `non-nullable Object`. If client declares `LoadingCache<String, Integer>`, then the type of `get()` would be non-nullable `Integer`. If the client declares `LoadingCache<String, @Nullable Integer>`, then the type of the getter becomes nullable.

By the way, as far as I remember, IDEA does not recognize `LoadingCache<K extends Object, V extends @Nullable Object>` syntax.

**Comment by vlsi** [comment]
However, `LoadingCache<K extends Object, V extends @Nullable Object> ` would conflict with `Cache<K extends Object, V extends Object>` that expects `non-nullable` `V`  :-/

**Comment by cpovirk** [comment]
I haven't studied the Caffeine API in enough detail to be 100% confident in what I'm about to say, but +1 to all of this:

> `LoadingCache` does not seem to have a relationship with its loader, so it looks like `@Nullable LoadingCache#get()` is the best possible type. How would `LoadingCache` know the loader never returns `null`? For instance, there might be an extra type parameter like `LoadingCache<K, V, LOADER_RETURN_TYPE>` just for the purpose of capturing nullness of the loader

Here are a couple other alternatives, each of which has its own downsides:

#### `getNonNull`

Give `LoadingCache` a new method, `default V getNonNull(K key)`. That would presumably perform a runtime null check on the result. That check guarantees that the annotation on the return type is correct. But of course it also means that calls to that method can throw, and static analyzers won't have a way to detect when that might happen (unless they have access to the code that builds the cache and they understand it -- very unlikely in practice).

#### `NonNullLoadingCache`

Define a subinterface `NonNullLoadingCache` that overrides `get` to return plain `V`. Instances would be created by passing `Caffeine` an instance of a new subinterface, `NonNullCacheLoader`, which likewise overrides `load`, etc. to return plain `V`.

This solution, like @vlsi 's solutions, should provide nullness checkers enough information to perform their checks.

But there are downsides:

- It is a step toward an exponential explosion of interfaces: Because it is orthogonal to `Async`-ness, we also end up with `NonNullAsyncLoadingCache` and `NonNullAsyncCacheLoader`. (Or is that `AsyncNonNullLoadingCache` and `AsyncNonNullCacheLoader`? :)) Fortunately, I think it's _not_ orthogonal to `Loading`-ness, since nullness is in question only for loading caches?
- It interacts weirdly with type inference: For a call like `build(key -> createExpensiveGraph(key))`, Java will always pick the more specific `NonNullCacheLoader` overload, even if `createExpensiveGraph` can return `null`. That means two things:
   - At best, nullness checkers will issue a compile error for loaders that can return `null`, forcing users to add a cast or extract a variable to control the type.
   - At worst, code will fail at runtime because the Kotlin compiler (and maybe other tools, but mainly Kotlin) will automatically insert runtime null checks on the result of each call to `NonNullLoadingCache.get`.

(To avoid the overloading / type-inference issues, I was going to suggest using a different name than the current `build` and `buildAsync`. But that's not an ideal solution for the Kotlin case: Lots of users will still call `build` and `buildAsync`, and they won't learn that they're "wrong" until Kotlin code throws an exception. Then again, if we can get to the point where the problem exists only for people who uses their caches from Kotlin but create them in Java and also don't use Java nullness checker, maybe we've won, relatively speaking?)

**Comment by cpovirk** [comment]
Oops, one more thought:

> However, `LoadingCache<K extends Object, V extends @Nullable Object> ` would conflict with `Cache<K extends Object, V extends Object>` that expects `non-nullable` `V` :-/

Yes, that's unfortunate.

One possibility is to generalize `Cache` to permit nullable types for `V`. We could do that without actually supporting such caches in practice: The generics on the `Caffeine` factory class could prevent creating such caches. Still, that is surprising at best. _Plus_, more thorough nullness checkers like the Checker Framework then won't like that `Cache.asMap()` returns a `ConcurrentMap`, as it's generally expected that `ConcurrentMap` implementations don't support nullable types. Similarly, the nullable `V` type may create complications for others methods: For example, `getAll`, _if I understand correctly_, will return a map that never contains nulls. But that information would be lost in a `LoadingCache<Foo, @Nullable Bar>`.

That all leads back to `LoadingCache<K, V, LOADER_RETURN_TYPE>`. That may ultimately have been too awkward no matter what, but it would have been a more interesting option if we'd had this conversation pre-3.0.0 :)

**Comment by cpovirk** [comment]
> * At worst, code will fail at runtime because the Kotlin compiler (and maybe other tools, but mainly Kotlin) will automatically insert runtime null checks on the result of each call to `NonNullLoadingCache.get`.

Sorry, there is a legitimate concern behind what I said, but what I said above is an overstatement.

If you merely removed `@Nullable` from the return type of `LoadingCache.get`, Kotlin wouldn't automatically insert null checks on the result of all calls. That would happen only if you explicitly annotated the method with a `@NonNull` annotation or annotated the scope with [a custom jsr305 defaulting annotation](https://kotlinlang.org/docs/java-interop.html#type-qualifier-defaults) like [this temporary one in Guava](https://github.com/google/guava/blob/cd3b4197fb523bf26f9122492915006ad306b3f3/guava/src/com/google/common/base/ElementTypesAreNonnullByDefault.java) -- or, someday, with our JSpecify [`@NullMarked`](https://jspecify.dev/user-guide.html#nullmarked) annotation. In particular, Kotlin does _not_ respect [the Checker Framework `@DefaultQualifier` annotations](https://github.com/ben-manes/caffeine/blob/15330a16b509b56ae32a9bd0cb703e36ff6caa1f/caffeine/src/main/java/com/github/benmanes/caffeine/cache/package-info.java#L37-L39). Thus, it would treat an unannotated `V` return type as a [platform type](https://kotlinlang.org/docs/java-interop.html#null-safety-and-platform-types).

(I still think the recommendations in my post and vlsi's are the ones I'd make.)

**Comment by ben-manes** [comment]
sorry for my tardy reply, but this week is hectic for me. Thanks for the excellent replies and I apologize if I haven't read them thoroughly enough in my response.

I don't think we would have wanted the complex generic signature even without backwards compatibility concerns, as it is not user friendly. I'd also argue that new interfaces, etc. just for nullness does not carry its conceptual weight. Both are good thought experiments for finding something that works, and then walking back to the best fit.

If Kotlin, et al. don't inject bytecode checks then that's excellent. It sounds like the right solution is to have the method's nullness be unspecified and assumed non-null for static analysis. This would match the common case usages while being clear in the JavaDoc. **If I understand correctly, the consensus is that removing the `@Nullable` is the right resolution here?**

---

For background, Guava's `LoadingCache.get(key)` throws an exception for a null value. At the time we were split: Kevin Bourrillion / Josh Bloch against null returns with checked exceptions, Charles Fry and I for it with runtime exception, and Josh felt strongly so we conceded (as the obvious expert). The argument against was to be null hostile throughout Guava, while we argued that the data may not exist in the external store (at no fault of the programmer) and be handled as normal control flow. The Guava choice did not seem be an exceptional condition (EJ Item 69-70) and the exception handling is verbose for this type control flow (`getUnchecked` was added later by popular request). Even back then we were considering `Optional<V>`, but storing an optional for negative caching would make the return type distasteful, there's no `flatten` method for nested cases (see Scala's), and the type still does not quite fit the language.

Doug Lea's addition of Map computations in Java 8 use null to indicate absence and removal if an existing mapping. As per Guava's intent to feel like a natural extension to the Collection Framework, Caffeine's refreshed apis try to align with what developers should be most familiar with. This also means that we differ by allowing `CacheLoader.reload` to return null to indicate that the entry should be removed, e.g. its should be viewed as an asynchronous `Map.compute`. Therefore `Cache.get` & `LoadingCache.get` methods merely a delegate to `computeIfAbsent`, exceptions are rethrown unchecked, and null is an expected response if non-computable.

**Comment by cpovirk** [comment]
The JSpecify position is something like: "The return type should be `@Nullable`, but that's inconvenient, and the JSpecify annotations and various tools need to do more work to make it more convenient."

At some point in the future, I would like us to live in a world in which you can have `@Nullable` on the return type: There _are_ cases in which the caller needs for the method to return `null`, so, if the question is "Is the type nullable are not?" then the answer is "Yes, it is."

One of our open questions for JSpecify is how we can make that `@Nullable` annotation less painful: Currently, we anticipate that people will do just what has been proposed here -- choose to omit `@Nullable` simply because the resulting API is more convenient in practice. And as long as they don't mark the return type as _non_-null, they're not saying anything _wrong_, just _incomplete_.

We have various ideas for reducing that pain, but since we can't change the language to add something like Kotlin's [`!!` operator](https://kotlinlang.org/docs/null-safety.html#the-operator), there's no one obvious answer.

So `@Nullable` is the goal, but we understand that we need to do more than say that in order to get people to actually do it :)

**Comment by DavyLandman** [comment]
bit late to the party, but I think this is a good solution. Any plans on a new release for this change? We are finally switching our language to use caffeine 3, and are running into this definition of LoadingCache as well ;)

**Comment by ben-manes** [comment]
Sorry for dragging my feet here. This is released in v3.0.5.
